File: generated-code/php/codecraft/TcpReadWrite/Model/AttackAction.php
<?php

namespace Model {
    require_once 'Model/AutoAttack.php';

    /**
     * Attack action
     */
    class AttackAction
    {
        /**
         * If specified, target entity's ID
         */
        public $target;
        /**
         * If specified, configures auto attacking
         */
        public $autoAttack;
    
        function __construct($target, $autoAttack)
        {
            $this->target = $target;
            $this->autoAttack = $autoAttack;
        }
    
        /**
         * Read AttackAction from input stream
         */
        public static function readFrom($stream)
        {
            if ($stream->readBool()) {
                $target = $stream->readInt32();
            } else {
                $target = NULL;
            }
            if ($stream->readBool()) {
                $autoAttack = \Model\AutoAttack::readFrom($stream);
            } else {
                $autoAttack = NULL;
            }
            return new AttackAction($target, $autoAttack);
        }
    
        /**
         * Write AttackAction to output stream
         */
        public function writeTo($stream)
        {
            if (is_null($this->target)) {
                $stream->writeBool(false);
            } else {
                $stream->writeBool(true);
                $stream->writeInt32($this->target);
            }
            if (is_null($this->autoAttack)) {
                $stream->writeBool(false);
            } else {
                $stream->writeBool(true);
                $this->autoAttack->writeTo($stream);
            }
        }
    }
}

File: generated-code/pascal/codecraft/TcpReadWrite/Model/AttackAction.pas
unit AttackAction;

interface

uses
    AutoAttack in 'Model/AutoAttack.pas',
    Nullable,
    Stream,
    SysUtils;

type
    // Attack action
    TAttackAction = class
        // If specified, target entity's ID
        target: TNullable<Int32>;
        // If specified, configures auto attacking
        autoAttack: TNullable<TAutoAttack>;
        constructor Create(target: TNullable<Int32>; autoAttack: TNullable<TAutoAttack>);
        // Read AttackAction from input stream
        class function ReadFrom(stream: TStream): TAttackAction; static;
        // Write AttackAction to output stream
        procedure WriteTo(stream: TStream);
        function ToString: ansistring; override;
    end;

implementation

constructor TAttackAction.Create(target: TNullable<Int32>; autoAttack: TNullable<TAutoAttack>);
begin
    self.target := target;
    self.autoAttack := autoAttack;
end;

class function TAttackAction.ReadFrom(stream: TStream): TAttackAction;
var autoAttack: TNullable<TAutoAttack>;
var autoAttackValue: TAutoAttack;
var target: TNullable<Int32>;
var targetValue: Int32;
begin
    if stream.ReadBoolean then begin
        targetValue := stream.ReadInt32;
        target := targetValue;
    end else
        target := nil;
    if stream.ReadBoolean then begin
        autoAttackValue := TAutoAttack.ReadFrom(stream);
        autoAttack := autoAttackValue;
    end else
        autoAttack := nil;
    result := TAttackAction.Create(target, autoAttack);
end;

procedure TAttackAction.WriteTo(stream: TStream);
var autoAttackValue: TAutoAttack;
var targetValue: Int32;
begin
    if target.HasValue then begin
        stream.WriteBoolean(true);
        targetValue := target.Value;
        stream.WriteInt32(targetValue);
    end else
        stream.WriteBoolean(false);
    if autoAttack.HasValue then begin
        stream.WriteBoolean(true);
        autoAttackValue := autoAttack.Value;
        autoAttackValue.WriteTo(stream);
    end else
        stream.WriteBoolean(false);
end;

function TAttackAction.ToString: ansistring;
var autoAttackValue: TAutoAttack;
var targetValue: Int32;
begin
    result := 'AttackAction {';
    result += 'target=';
    if target.HasValue then begin
        targetValue := target.Value;
        result += IntToStr(targetValue);
    end else
        result += 'nil';
    result += ', ';  
    result += 'autoAttack=';
    if autoAttack.HasValue then begin
        autoAttackValue := autoAttack.Value;
        result += autoAttackValue.ToString;
    end else
        result += 'nil';
    result += '}';
end;

end.